ENH: Add annotations for five array construction functions

BvB93 opened the pull request:
This pull request adds annotations to the five array construction functions in `np.core._asarray`:

* `asarray`
* `asanyarray`
* `ascontiguousarray`
* `asfortranarray`
* `require`

Commit:
ENH,WIP: Added type hints to `np.core._asarray`

Commit:
ENH: Use literals where possible

Commit:
TST: Added tests for `np.core._asarray`

Commit:
TST: Consolidate all array construction tests

Commit:
ENH: Update the `np.__all__`

Commit:
MAINT: Minor fixes; removed unused imports

BvB93 commented on numpy/core/_asarray.pyi:
```diff
@@ -0,0 +1,76 @@
+) -> ndarray: ...
+@overload
+def asanyarray(
+    a: _ArrayType,
```
`def(a: _ArrayType, ...) -> _ArrayType` implies that both the array (sub-)class, dtype and shape are identical for the input and output.

Without higher-kindred typevar we, unfortunately, can not describe a situation were an `ndarray` subclass is returned with a dtype that is different from the initial input (hence the `dtype: None` in this overload).

BvB93 commented on numpy/core/_asarray.pyi:
```diff
@@ -0,0 +1,76 @@
+def require(
+    a: object,
+    dtype: DtypeLike = ...,
+    requirements: Union[_E, Iterable[Union[_E, _Requirements]]] = ...,
```
`Iterable[Union[_E, _Requirements]]`: an iterable which contains an element from `E` and may or may not contain additional elements from  `_Requirements`.

eric-wieser commented on numpy/core/_asarray.pyi:
```diff
@@ -0,0 +1,76 @@
+    "W", "WRITEABLE",
+    "O", "OWNDATA"
+]
+_E = Literal["E", "ENSUREARRAY"]
```
Perhaps
```suggestion
_E = Literal["E", "ENSUREARRAY"]
_Requirements_With_E = Union[_Requirements, _E]
```

BvB93 replied:
Good idea; done in https://github.com/numpy/numpy/pull/17326/commits/07476cfb144baf9f6a889076d02b6e80261e3c76.

Commit:
ENH: Add an alias for `Union[_Requirements, _E]`

Addresses https://github.com/numpy/numpy/pull/17326#discussion_r489322146

BvB93 commented:
Rebased to get rid of a merge conflict.

Commit:
MAINT: Replaced `__all__` with `import x as x`

Objects defined in an external file will be picked up if they are either specified in `__all__` or imported as `import x as x`.
The latter notation is more compact and is less prone to merge conflicts.

BvB93 commented:
https://github.com/numpy/numpy/pull/17326/commits/71def2cf787a592c72484ae676fac353b14432d7 implements an alternative approach for ensuring that externally-defined stubs are 
picked up in the main `__init__.pyi` file. Using `from x import y as y` is more compact than 
specifying them in `__all__` and, more importantly, it is less prone to merge conflicts.

Commit:
MAINT: Removed placeholder annotations

BvB93 commented:
https://github.com/numpy/numpy/pull/17326/commits/be77442c1facae452d3742415161c4f8203f4a86 removes the placeholder annotations (see https://github.com/numpy/numpy/pull/17406) for the newly annotated functions introduced in this PR. 

Secondly, if there are no further comments then I feel this can get merged.

charris commented:
Thanks @BvB93 .
